Add feature image cropper for upload

Ddharmani3 opened the pull request:
added image editor from uppy for this feature
this will allow users to custom crop their images at upload

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
#24324
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://user-images.githubusercontent.com/64723994/220267775-ef8accd4-29b3-4d78-8064-a8dac671d14c.mp4

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! As noted in the PR template, please add still screenshots of the changes as well.

Ddharmani3 commented:
> Thanks! As noted in the PR template, please add still screenshots of the changes as well.

@alya done! added the screenshots as well.
waiting for your review to complete the task.

timabbott commented on static/js/upload_widget.ts:
```diff
@@ -45,11 +134,30 @@ export function build_widget(
+        $clear_button.hide();
+        $(
+            "#add-custom-emoji-modal .dialog_submit_button, #add-custom-emoji-modal .dialog_cancel_button,.emoji_name_input",
+        ).show();
```
This selector is pretty weird -- we should probably want to compute what to show from which element we are.

(And I don't think we want to hardcode this image cropping feature to just work with the custom emoji input -- we should make it an optional setting for the upload widget).

Ddharmani3 replied:
Yes, I am aware that hardcoding the features is not recommended, and I will make sure to implement them as soft-coded when working on the remaining part of the issue, which includes adding the same feature to profile pictures and organization logo uploads.

timabbott commented on static/js/upload_widget.ts:
```diff
@@ -25,17 +39,92 @@ export function build_widget(
     $upload_button: JQuery,
     $preview_text?: JQuery,
     $preview_image?: JQuery,
+    $select_button?: JQuery,
```
Why is this called `$select_button`?

Ddharmani3 replied:
$select_button refers to the button that is used to select the cropped image, which is why I named it as such. However, I am open to suggestions for alternative names. If you have any other ideas for what to name the button, I would be happy to hear them and consider making the change.

timabbott commented on static/js/settings_emoji.js:
```diff
@@ -248,9 +252,7 @@ function show_modal() {
-        for (const [i, file] of Array.prototype.entries.call($("#emoji_file_input")[0].files)) {
-            formData.append("file-" + i, file);
-        }
+        formData.append("file-0", upload_widget.get_cropped_file());
```
This seems like it'll prevent uploading multiple images at once in the emoji upload UI -- is that expected?

Ddharmani3 replied:
Yes, it is expected that only one image can be uploaded. Even before I added the cropping feature, the upload widget was restricted to only allow one image. I have maintained the same restriction with uppy as well, so it won't have any impact.

timabbott commented:
@Ddharmani3 thanks for working on this! Can you talk about the technical design choices here -- I notice you use 3 different Uppy plugins; can you explain what they're for?

Ddharmani3 commented:
> @Ddharmani3 thanks for working on this! Can you talk about the technical design choices here -- I notice you use 3 different Uppy plugins; can you explain what they're for?

@timabbott  I utilized three uppy plugins because the uppy image editor is designed to be used with uppy dashboard (as mentioned [here](https://uppy.io/docs/image-editor/)). However, uppy dashboard requires uppy core to function properly, so all three plugins are needed in order to use uppy image editor."

Ddharmani3 commented:
@timabbott, I have addressed all of your inquiries in my previous responses. If you have any further questions or concerns, please bring them up to me, and I will surely work on them. However, if you are satisfied with the current design and features, I will proceed with implementing the feature for other image uploads. I will ensure that the upload widget is soft-coded (as previously mentioned) and I am open to changing the $select_button name to a name of your choice.

Ddharmani3 commented:
@timabbott, I have incorporated an image editor into all uploads. Please review the PR at your convenience.

However, I must mention that there is one issue with eslint that I am unable to resolve due to a cyclic dependency. Specifically, upload_widget.ts is calling dialog_widget, which is calling ./overlays:8=>./popovers:6=>./settings_bots:38=>./avatar:11=>, which in turn is calling the upload_widget itself.

I have requested for a fix for this issue in [CZO](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/import.20dependency.20cycle/near/1517532). Till then, I have added "// eslint-disable-next-line import/no-cycle" in upload_widget.ts to ignore this issue.

If anyone reviewing this PR knows how to fix this issue, I would greatly appreciate any guidance or assistance you can offer. Thank you in advance.

Ddharmani3 commented:
@timabbott since we are using uppy for image editing at upload time. I think it's possible to explore other Uppy upload features that could be very useful for users. For example, in addition to the current option of selecting files from the device, we could also provide drag & drop, Instagram and Facebook upload options that would be helpful for users to upload their profile pictures directly from their social handles. We could also provide drive uploads that would be useful for organization icons/logos directly from drives. What do you think about this idea?

timabbott commented on tsconfig.json:
```diff
@@ -1,5 +1,7 @@
 {
     "compilerOptions": {
+        "skipLibCheck": true,
+        "allowJs": true,
```
Why do we need these changes?

Ddharmani3 replied:
I am using `allowjs` to import `dialog_widget` in `upload_widget.ts` since it is a JavaScript module, and `skiplibcheck` to suppress errors that arise due to using `allowjs`. However, once dialog_widget is converted to TypeScript, these options will no longer be necessary. I have requested a solution on [CZO](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/use.20javascript.20module.20in.20typescript/near/1533464) and will update the PR once I receive one.

timabbott commented on web/src/upload_widget.ts:
```diff
@@ -25,31 +41,122 @@ export function build_widget(
+                        cropWidescreen: false,
+                        cropWidescreenVertical: false,
+                    },
+                })
```
Is there a way to avoid this large block of duplicated uppy configuration code? You seem to do this twice; once here and once in `accept` below.

Ddharmani3 replied:
I have addressed this issue in the updated PR and have reduced the redundancy.

timabbott commented:
@Ddharmani3 can you reply to the above review feedback? I believe we merged the import cycle fixes that were blocking this.

Ddharmani3 commented:
> @Ddharmani3 can you reply to the above review feedback? I believe we merged the import cycle fixes that were blocking this.
@timabbott, I apologize for my delayed response to your feedback. I was hoping to address the issues by pushing new changes, but my branch got messed up due to rebasing. I asked for help on [CZO](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/typescript_type_error/near/1523775), but I found out today that all the issues were caused by using the js module "dialog_widget" in "upload_widget.ts". I have fixed everything else, but this issue still needs to be resolved. I believe it can be fixed by converting "dialog_widget" to TypeScript, but until then, I have asked for a solution on [CZO](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/use.20javascript.20module.20in.20typescript/near/1533464).

I'd also like to bring up another issue with Uppy's image editor. Currently, it doesn't support editing GIFs; it converts them to static images instead. In my PR, I've handled this by displaying the text "GIFs cannot be edited yet" in the input error div for emoji, bot avatar uploads, and not opening the Uppy image editor. However, for profile picture/organization logos and icons, I do open the Uppy editor, which allows users to crop GIFs and convert them to static images, as these areas don't support GIFs anyway. I have asked for a plugin that supports GIF cropping [here](https://community.transloadit.com/t/gif-editing-support/16510/3?u=dharmanihardik). What are your thoughts on this handling?

timabbott commented:
Would it be possible to just detect whether the image is an animated image, and if it is, skip the image editor step, without showing an error?

Ddharmani3 commented:
> Would it be possible to just detect whether the image is an animated image and if it is, skip the image editor step, without showing an error?

Hi @timabbott, I am doing exactly the same, for all other uploads except custom_emoji upload, I am opening the uppy image editor and letting users edit GIFs. Since these fields do not show animated images anyway, it's fine even if they are converted to a static image. This behavior is the same as before where users could upload GIFs but they were converted to static images.

https://user-images.githubusercontent.com/64723994/229377495-6383c54c-d4bf-40e4-ae34-e82a0278ef68.mp4

However, for custom_emoji uploads, they support showing animated GIFs. In that case, I skip the image editor and just show a text in grey color that says 'GIF cannot be edited yet' so users don't get confused about the image editor not opening.

https://user-images.githubusercontent.com/64723994/229377507-0aead9c3-9e0e-4826-8970-5a41e0cdff98.mp4

Since dialog_widget is converted to typescript now, I have pushed the updated commits. This PR is ready for review now.

Ddharmani3 commented:
@timabbott, I have resolved conflicts. This PR is ready for your review now.

Ddharmani3 commented:
@timabbott can we add this feature in the 7.0 release?

alya commented:
We are not planning to specifically prioritize this for the 7.0 release.

Note that Zulip Cloud users see features within a week or two after they are merged, so your work won't be sitting on a shelf in any case. :)

alya commented:
In testing this PR (profile picture upload, specifically), I'm seeing the image shift around seemingly randomly, maybe as a result of me moving the mouse. (I'm not clicking on anything.)

Ddharmani3 commented:
> In testing this PR (profile picture upload, specifically), I'm seeing the image shift around seemingly randomly, maybe as a result of me moving the mouse. (I'm not clicking on anything.)
> 
> Maybe it's just zooming? But it still shouldn't do that.

Are you mentioning this zoom behavior: 

https://user-images.githubusercontent.com/64723994/232237467-74e56271-ae91-432d-9e7e-67493489f5c0.mp4

Also, I want feedback on this:
![image](https://user-images.githubusercontent.com/64723994/232237525-a340ce40-0b94-4462-90b4-86072be7c247.png)

The first[revert] and the last[cropsquare] options can change the selected Aspect ratio. The `revert` option makes the aspect ratio to be equal to the image's aspect ratio and the `crop square` option makes the aspect ratio 1:1 (square) which is not a problem for almost all the uploads except the organization logo which has an 8:1 aspect ratio.

Should I remove these options? Like this: 
![image](https://user-images.githubusercontent.com/64723994/232238104-3d4a9cc4-4d2b-48ab-bcf8-d9d16b7e6f86.png)

alya commented:
> Are you mentioning this zoom behavior:

I don't know how to tell what you were doing to cause the behavior from the video, but probably? Yes, please remove it.

alya commented:
Hm, we should prevent the user from accidentally ending up with the wrong aspect ratio. I guess we can keep the `crop square` option just for the square image uploads, and drop the `revert` option?

Ddharmani3 commented:
> > Are you mentioning this zoom behavior:
> 
> I don't know how to tell what you were doing to cause the behavior from the video, but probably? Yes, please remove what you have in mind, and I'll retest to see if the widget feels hard to use after that.

@alya I have removed the zoom feature on mouse touch/scroll but the image still can be `zoomed in/zoomed out` by selecting the options in the bar if user wants.

Also, removed the revert option for all uploads and kept crop square for all uploads except the organization logo upload.

These changes are ready for review now.

alya commented:
I'm seeing the upload window close unexpectedly when I drag the selection box and release the mouse outside the modal.


![2023-04-25 14 29 59](https://user-images.githubusercontent.com/2090066/234408589-8d9d366b-c312-4d0e-8bb4-89d1c14ef44c.gif)


I think you need to really play with this one to make sure there aren't any interactions that feel broken.

Ddharmani3 commented:
@zulipbot add "QA needed"

Ddharmani3 commented:
Added "QA needed" label so if anyone is interested in doing manual testing for this new feature to look for any UI/UX issues like the one found by @alya in the above comment, fixed that issue in the 3rd commit. I will also try to play more with this new feature but it would be great if folks can help find unnoticed issues :)

Ddharmani3 commented:
@zulipbot add "buddy review"

prakhar1144 reviewed (commented):
### Bug 1
<details><summary>Minimize the cropper to its full capacity -- See top-left</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/8a198c27-af61-4abe-918a-8c068734d0ce">
</img>
</details> 
<details><summary>Error</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/6a5b7d17-a642-4bc9-8b4b-d2952ac7c162">
</img>
</details> 

### Bug 2
<details><summary>Select an area outside image -- See top-left</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/dfcc4adb-7d29-42d7-a24c-ab92f22de316">
</img>
</details> 
<details><summary>Blank image saved</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/d3deeca5-4244-47dd-b4f1-7ef0d10b5803">
</img>
</details> 

### Question
<details><summary>This pop-up appears for a fraction of second after we click on "Crop & Select" -- can / should we remove this?</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/24018baf-6671-4f5c-b74a-1a2db436d415">
</img>
</details>

Ddharmani3 commented:
@prakhar1144  I have resolved the conflicts and here is the status of bug fixing:
Bug 2:
There is a possibility of having a blank image in the current upload modal in Zulip/main too if the user uploads an empty image. However, I am aware that in the Image editor, if a user mistakenly selects a crop area outside the image due to a misclick, it might be an issue/ bad experience. But, I am not sure how to check if the selected area is completely transparent. I will try to find a way. If you know any approach do share it with me and I would love to work on it :). However even if we can't fix it, I don't believe this can be considered a bug since users can always re-upload the image if it is mis-cropped.

Bug 3/ Question: 
The Uppy Image Editor is officially intended to be used only with the Uppy Dashboard. The screen you mentioned is the Uppy Dashboard, and there is an option, `autoOpenFileEditor` which I set to true so that the dashboard is not initially displayed and the Image editor is opened directly. I have implemented custom logic to automatically close the dashboard upon completion of the crop, as there is no default option provided to auto-close the dashboard after completing the crop and hence with my logic it appears for a fraction of a second after closing as it is not supported by default with uppy. Without `autoOpenFileEditor: true`, the flow would appear as follows:

https://github.com/zulip/zulip/assets/64723994/e30ece14-49fe-4610-87bb-3c18188d16ff


The screen at the start is uppy dashboard and by clicking on the pencil edit icon Image editor opens; But, with  `autoOpenFileEditor: true` it directly opens the Image editor but there is no similar option to auto close by default.

Bug 1: I think this is definitely a very important bug to solve. But I am unable to reproduce it, I have tried several times and no matter how small I make the crop area I wasn't able to reproduce it :(. Can you share if you have found any specific way to reproduce this accurately? I will try it again tomorrow to reproduce it.

prakhar1144 commented:
> Bug 1: I think this is definitely a very important bug to solve. But I am unable to reproduce it, I have tried several times and no matter how small I make the crop area I wasn't able to reproduce it

[zulip-error.webm](https://github.com/zulip/zulip/assets/56781761/beca4855-b5ae-415e-8cf5-a1d021c9577c)
